### ISSUE TITLE
Dendrite with Cloudflare Tunnel fails federation test

### ISSUE THREAD

**Issue by gBasil** [matrix-org/dendrite#2971]
### Background information

- **Dendrite version or git SHA**: 0.11.1+cc59879
- **SQLite3 or Postgres?**: Postgres
- **Running in Docker?**: Yes
- **`go version`**: 1.20
- **Client used (if applicable)**: Element Web

### Description

I set up Dendrite (forwarding the 8008 port), and Caddy (serving Element Web) in a Docker Compose setup. My `Caddyfile` is as follows:
```
# http:// is used to fix some weird certificate issue with Cloudflare. It doesn't work without it.
http://matrix.<domain> {
	handle /.well-known/matrix/server {
		header Content-Type application/json
		header Access-Control-Allow-Origin *
		respond `{"m.server": "matrix.<domain>"}`
	}

	handle /.well-known/matrix/client {
		header Content-Type application/json
		header Access-Control-Allow-Origin *
		respond `{"m.homeserver": {"base_url": "https://matrix.<domain>:443"}}`
	}

	# No idea which ones are needed
	reverse_proxy /_matrix/* :8008
	reverse_proxy /_synapse/* :8008

	root * /data/element
	file_server
}
```
`cloudflared` (the tunnel daemon) is running. Visiting matrix.<domain> displays the Element Web client, and logging in to the account I created works, so both Element and Dendrite seem to be working. However, when attempting to join a room (#gitea-federation-chat:matrix.org), I encountered an M_UNKNOWN error. Checking the logs gives me this message:
```
time="2023-02-15T06:21:46.264165239Z" level=error msg="federation.LookupRoomAlias failed" error="contents=[<redacted for brevity>] msg=Failed to GET JSON (hostname \"matrix.org\" path \"/_matrix/federation/v1/query/directory\") code=401 wrapped=M_UNAUTHORIZED: Failed to find any key to satisfy: _FetchKeyRequest(server_name='matrix.<redacted>', minimum_valid_until_ts=1676442096182, key_ids=['ed25519:<redacted>'])" req.id=<redacted> req.method=GET req.path="/_matrix/client/r0/directory/room/#gitea-federation-chat:matrix.org
```
When running the Federation Tester (https://federationtester.matrix.org), the API returns the following:
```json
{
  "WellKnownResult": {
    "m.server": "",
    "result": "No .well-known found",
    "CacheExpiresAt": 0
  },
  "DNSResult": {
    "SRVSkipped": false,
    "SRVCName": "",
    "SRVRecords": null,
    "SRVError": {
      "Message": "lookup _matrix._tcp.matrix.<redacted> on 8.8.8.8:53: no such host"
    },
    "Hosts": {
      "matrix.<redacted>": {
        "CName": "matrix.<redacted>.",
        "Addrs": [
          "2606:4700:3031::6815:39d5",
          "2606:4700:3031::ac43:a708",
          "104.21.57.213",
          "172.67.167.8"
        ],
        "Error": null
      }
    },
    "Addrs": [
      "[2606:4700:3031::6815:39d5]:8448",
      "[2606:4700:3031::ac43:a708]:8448",
      "104.21.57.213:8448",
      "172.67.167.8:8448"
    ]
  },
  "ConnectionReports": {},
  "ConnectionErrors": {
    "104.21.57.213:8448": {
      "Message": "Get \"https://104.21.57.213:8448/_matrix/key/v2/server\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
    },
    "172.67.167.8:8448": {
      "Message": "Get \"https://172.67.167.8:8448/_matrix/key/v2/server\": dial tcp 172.67.167.8:8448: i/o timeout (Client.Timeout exceeded while awaiting headers)"
    },
    "[2606:4700:3031::6815:39d5]:8448": {
      "Message": "Get \"https://[2606:4700:3031::6815:39d5]:8448/_matrix/key/v2/server\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
    },
    "[2606:4700:3031::ac43:a708]:8448": {
      "Message": "Get \"https://[2606:4700:3031::ac43:a708]:8448/_matrix/key/v2/server\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
    }
  },
  "Version": {
    "error": "Get \"matrix://matrix.<redacted>/_matrix/federation/v1/version\": dial tcp [2606:4700:3031::ac43:a708]:8448: i/o timeout"
  },
  "FederationOK": false
}
```

- **What** is the problem: Dendrite failing to connect to a room, and the Federation Tester test fails
- **Who** is affected: Server admin (and it's only user, the admin)
- **How** is this bug manifesting: Error message when trying to join a room and when running Federation Tester
- **When** did this first appear: N/A, this is a new server, so it couldn't have worked before

I'm aware of the argument of not using Cloudflare due to privacy concerns, but port forwarding is not an option for me.

### Steps to reproduce

- See above.

Thank you for your time! Sorry if this is a beginners mistake.

**Comment by spaetz** [comment]
This is no place for user support but a couple observations. I don't know caddy, but you configure
http://marix.* (not https).

- you need https, no http connections.
- You have no .well-known on your main domain and no SRV DNS entry, so federationtester correctly tries port 8448 for the S2S connection, which you don't serve.
- cloudflare is weird in many regards, and while people seem to run with it just fine, it might pose another hurdle beyond the 2 above.

**Comment by gBasil** [comment]
@spaetz 
> This is no place for user support
Sorry! 😅 I didn't want to register an account on matrix.org only to delete it once I was done.

> you need https, no http connections.
Cloudflare already configures HTTPS. When I tried having Caddy configure it's own (or even a CF Origin certificate), requests to the subdomain refused to connect.

> You have no .well-known on your main DOMAIN
Oh! That flew by me, didn't know it was on the main domain. Thanks!

I found my solution on a blog post: https://www.frrobert.net/x/gemini.macdermid.ca/matrix.gmi, which was adding a port (`443`) to the `m.server` value.

Thank you for your time, and sorry for the issue.
